<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>MOP_MOD &mdash; MoP2  documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
    <link rel="canonical" href="https://biocorecrg.github.io/MoP2/docs/mop_mod.html" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/sphinx_highlight.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="author" title="About these documents" href="about.html" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="MOP_CONSENSUS" href="mop_consensus.html" />
    <link rel="prev" title="MOP_TAIL" href="mop_tail.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="index.html" class="icon icon-home"> MoP2
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="about.html">About Master of Pores</a></li>
<li class="toctree-l1"><a class="reference internal" href="install.html">How to install</a></li>
<li class="toctree-l1"><a class="reference internal" href="mop_preprocess.html">MOP_PREPROCESS</a></li>
<li class="toctree-l1"><a class="reference internal" href="mop_tail.html">MOP_TAIL</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">MOP_MOD</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#input-parameters">Input Parameters</a></li>
<li class="toctree-l2"><a class="reference internal" href="#how-to-run-the-pipeline">How to run the pipeline</a></li>
<li class="toctree-l2"><a class="reference internal" href="#results">Results</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="mop_consensus.html">MOP_CONSENSUS</a></li>
<li class="toctree-l1"><a class="reference internal" href="reporting.html">Resource reporting</a></li>
<li class="toctree-l1"><a class="reference internal" href="benchmark.html">Benchmark</a></li>
<li class="toctree-l1"><a class="reference internal" href="changelog.html">CHANGELOG</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">MoP2</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
      <li>MOP_MOD</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/mop_mod.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="mop-mod">
<span id="home-page-mopmod"></span><h1>MOP_MOD<a class="headerlink" href="#mop-mod" title="Permalink to this heading"></a></h1>
<p>This module takes as input the output from MOP_PREPROCESS: basecalled fast5 reads, together with their respective fastq files and unspliced alignments to the transcriptome . It runs four different RNA detection algorithms (Epinano, Nanopolish, Tombo and Nanocompore) and it outputs the predictions generated by each one of them as individual tab-delimited files.</p>
<section id="input-parameters">
<h2>Input Parameters<a class="headerlink" href="#input-parameters" title="Permalink to this heading"></a></h2>
<table class="colwidths-given docutils align-default">
<colgroup>
<col style="width: 25%" />
<col style="width: 75%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Parameter name</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><strong>input_path</strong></p></td>
<td><p>Output folder generated by mop_preprocess</p></td>
</tr>
<tr class="row-odd"><td><p><strong>comparison</strong></p></td>
<td><p>TSV file with two fields, each one will indicate the ID of the sample that has to be compared 1 vs 1</p></td>
</tr>
<tr class="row-even"><td><p><strong>reference</strong></p></td>
<td><p>reference sequences</p></td>
</tr>
<tr class="row-odd"><td><p><strong>output</strong></p></td>
<td><p>Output folder</p></td>
</tr>
<tr class="row-even"><td><p><strong>pars_tools</strong></p></td>
<td><p>TSV file with optional extra command line parameters for the tool indicated in the first field.</p></td>
</tr>
<tr class="row-odd"><td><p><strong>epinano</strong></p></td>
<td><p>It (in)activate the corresponding flow. It can be YES or NO</p></td>
</tr>
<tr class="row-even"><td><p><strong>nanocompore</strong></p></td>
<td><p>It (in)activate the corresponding flow. It can be YES or NO</p></td>
</tr>
<tr class="row-odd"><td><p><strong>tombo_lsc</strong></p></td>
<td><p>It (in)activate the corresponding flow. It can be YES or NO</p></td>
</tr>
<tr class="row-even"><td><p><strong>tombo_msc</strong></p></td>
<td><p>It (in)activate the corresponding flow. It can be YES or NO</p></td>
</tr>
<tr class="row-odd"><td><p><strong>epinano_plots</strong></p></td>
<td><p>If YES will produce a plot for each sample for each transcript.</p></td>
</tr>
<tr class="row-even"><td><p><strong>email</strong></p></td>
<td><p>Email for pipeline reporting.</p></td>
</tr>
</tbody>
</table>
</section>
<section id="how-to-run-the-pipeline">
<h2>How to run the pipeline<a class="headerlink" href="#how-to-run-the-pipeline" title="Permalink to this heading"></a></h2>
<p>Before launching the pipeline,user should:</p>
<ol class="arabic simple">
<li><p>Decide which containers to use - either docker or singularity <strong>[-with-docker / -with-singularity]</strong>.</p></li>
<li><p>Fill in both <strong>params.config</strong> and <strong>tools_opt.tsv</strong> files.</p></li>
<li><p>Fill in <strong>comparison.tsv</strong> file - please see example below:</p></li>
</ol>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">wt_1 ko_1</span>
<span class="go">wt_2 ko_2</span>
</pre></div>
</div>
<p>To launch the pipeline, please use the following command:</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">nextflow run mop_mod.nf -with-singularity &gt; log.txt</span>
</pre></div>
</div>
<p>You can run the pipeline in the background adding the nextflow parameter <strong>-bg</strong>:</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">nextflow run mop_mod.nf -with-singularity -bg &gt; log.txt</span>
</pre></div>
</div>
<p>You can change the parameters either by changing <strong>params.config</strong> file or by feeding the parameters via command line:</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">nextflow run mop_mod.nf -with-singularity -bg --output test2 &gt; log.txt</span>
</pre></div>
</div>
<p>You can specify a different working directory with temporary files:</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">nextflow run mop_mod.nf -with-singularity -bg -w /path/working_directory &gt; log.txt</span>
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<ul class="simple">
<li><p>In case of errors you can troubleshoot seeing the log file (log.txt) for more details. Furthermore, if more information is needed, you can also find the working directory of the process in the file. Then, access that directory indicated by the error output and check both the <strong>.command.log</strong> and <strong>.command.err</strong> files.</p></li>
</ul>
</div>
<div class="admonition tip">
<p class="admonition-title">Tip</p>
<p>Once the error has been solved or if you change a specific parameter, you can resume the execution with the <strong>Netxtlow</strong> parameter <strong>- resume</strong> (only one dash!). If there was an error, the pipeline will resume from the process that had the error and proceed with the rest. If a parameter was changed, only processes affected by this parameter will be re-run.</p>
</div>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">nextflow run mop_mod.nf -with-singularity -bg -resume &gt; log_resumed.txt</span>
</pre></div>
</div>
<p>To check whether the pipeline has been resumed properly, please check the log file. If previous correctly executed process are found as <em>Cached</em>, resume worked!</p>
</section>
<section id="results">
<h2>Results<a class="headerlink" href="#results" title="Permalink to this heading"></a></h2>
<p>Several folders are created by the pipeline within the output directory specified by the <strong>output</strong> parameter:</p>
<ol class="arabic simple">
<li><p><strong>Epinano</strong> results are stored in <strong>epinano_flow</strong> directory. It contains two files per sample: one containing data at position level and the other, at 5-mer level. Different features frequencies as well as quality data are included in the results. See example below:</p></li>
</ol>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">#</span>Ref,pos,base,cov,q_mean,q_median,q_std,mis,ins,del
<span class="go">gene_A,2515,C,45497.0,5.36995,4.00000,3.97797,0.0822032221904741,0.18715519704595907,0.2058377475437941</span>
<span class="go">gene_A,2516,A,45504.0,5.38207,4.00000,4.71619,0.17128164556962025,0.20497099156118143,0.07733386075949367</span>
<span class="go">gene_A,2517,C,45529.0,6.92130,5.00000,5.04250,0.06165301236574491,0.1505633771881658,0.13540820136616222</span>
<span class="go">gene_A,2518,A,45545.0,6.49821,5.00000,5.47485,0.10802503018992206,0.10855198155670216,0.2082775277198375</span>
<span class="go">gene_A,2519,T,45557.0,6.51247,5.00000,4.81853,0.09386043857145993,0.14792457800118533,0.2033057488421099</span>
</pre></div>
</div>
<p>Here an example of a plot from Epinano:</p>
<a class="reference internal image-reference" href="_images/epinano.png"><img alt="_images/epinano.png" src="_images/epinano.png" style="width: 600px;" /></a>
<ol class="arabic simple" start="2">
<li><p><strong>Tombo</strong> results are stored in <strong>tombo_flow</strong> directory. It contains one file per comparison. It reports the p-value per position, the sum of p-values per 5-mer and coverage in both WT and KO. See example below:</p></li>
</ol>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">&quot;Ref_Position&quot;       &quot;Chr&quot;   &quot;Position&quot;      &quot;Tombo_SiteScore&quot;       &quot;Coverage_Sample&quot;       &quot;Coverage_IVT&quot;  &quot;Tombo_KmerScore&quot;</span>
<span class="go">&quot;gene_A_3&quot;   &quot;gene_A&quot;        &quot;3&quot;     &quot;0.0000&quot;        &quot;92&quot;    &quot;87&quot;    NA</span>
<span class="go">&quot;gene_A_4&quot;   &quot;gene_A&quot;        &quot;4&quot;     &quot;0.0000&quot;        &quot;92&quot;    &quot;87&quot;    NA</span>
<span class="go">&quot;gene_A_5&quot;   &quot;gene_A&quot;        &quot;5&quot;     &quot;0.0000&quot;        &quot;92&quot;    &quot;87&quot;    0</span>
<span class="go">&quot;gene_A_6&quot;   &quot;gene_A&quot;        &quot;6&quot;     &quot;0.0000&quot;        &quot;93&quot;    &quot;88&quot;    0.0014</span>
<span class="go">&quot;gene_A_7&quot;   &quot;gene_A&quot;        &quot;7&quot;     &quot;0.0000&quot;        &quot;95&quot;    &quot;89&quot;    0.0027</span>
<span class="go">&quot;gene_A_8&quot;   &quot;gene_A&quot;        &quot;8&quot;     &quot;0.0014&quot;        &quot;95&quot;    &quot;89&quot;    0.004</span>
</pre></div>
</div>
<ol class="arabic simple" start="3">
<li><p><strong>Nanopolish</strong> results are stored in <strong>nanopolish-compore_flow</strong> directory. It contains two files per sample: raw eventalign output (gzipped) and another with the median raw current per position and transcript (<strong>sample_processed_perpos_median.tsv.gz</strong>). See example below:</p></li>
</ol>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">contig       position        reference_kmer  read_name       median  coverage</span>
<span class="go">gene_A       0       AAATT   1       113.35  433</span>
<span class="go">gene_A       1       AATTG   1       97.24   506</span>
<span class="go">gene_A       2       ATTGA   1       70.35   2034</span>
<span class="go">gene_A       3       TTGAA   1       102.03  416</span>
<span class="go">gene_A       4       TGAAG   1       115.315 422</span>
<span class="go">gene_A       5       GAAGA   1       104.25  471</span>
</pre></div>
</div>
<ol class="arabic simple" start="4">
<li><p><strong>Nanocompore</strong> results are stored in <strong>nanopolish-compore_flow</strong> directory. It contains one file per comparison (<strong>wt_1_vs_ko_1_nanocompore_results.tsv</strong>). Default output from Nanocompore (see Nanocompore’s repository for a more detailed explanation).</p></li>
</ol>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="mop_tail.html" class="btn btn-neutral float-left" title="MOP_TAIL" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="mop_consensus.html" class="btn btn-neutral float-right" title="MOP_CONSENSUS" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright .</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>